apps/bluetooth/js/transfer.js

```diff
@@ -195,36 +196,89 @@ navigator.mozL10n.once(function showPanel() {
-        var getRequest = storage.get(filepath);
+        var getRequest;
+        if (activity.source.data.filepaths) {
+          // There is filepath, apply approach 1.
+          // browse the file via filepath from storage again.
+          var filepath = activity.source.data.filepaths[index];
+          getRequest = storage.get(filepath);
+          getRequest.onsuccess = function() {
+            defaultAdapter.sendFile(targetDevice.address, getRequest.result);
+            var msg = 'getFile succeed & file is sending...';
+            debug(msg);
+            if (--numberOfTasks === 0) {
+              transferred();
+            }
+          };
 
-        getRequest.onsuccess = function() {
-          defaultAdapter.sendFile(targetDevice.address, getRequest.result);
-          var msg = 'getFile succeed & file is sending...';
+          getRequest.onerror = function() {
+            defaultAdapter.sendFile(targetDevice.address, blob);
+            var msg = 'getFile failed so that blob is sending without ' +
+                      'filename ' + getRequest.error && getRequest.error.name;
+            debug(msg);
+            if (--numberOfTasks === 0) {
+              transferred();
+            }
+          };
+        } else {
+          // There is no filepath, apply approach 2.
+          // Wrap the file with filename for sending.
+          // And the filename is better coming from the requsting app.
+          var TMPDIR = '.bluetooth/tmp';
+          var filename = TMPDIR + '/' + activity.source.data.filenames[index];
+          msg = 'add filename for blob, name = ' + filename;
           debug(msg);
-          if (--numberOfTasks === 0) {
-            transferred();
-          }
-        };
 
-        getRequest.onerror = function() {
-          defaultAdapter.sendFile(targetDevice.address, blob);
-          var msg = 'getFile failed so that blob is sending without filename ' +
-                    getRequest.error && getRequest.error.name;
-          debug(msg);
-          if (--numberOfTasks === 0) {
-            transferred();
-          }
-        };
+          var addNamedRequest = storage.addNamed(blob, filename);
+          addNamedRequest.onsuccess = function() {
+            var getRequest = storage.get(addNamedRequest.result);
+            getRequest.onsuccess = function() {
+              defaultAdapter.sendFile(targetDevice.address, getRequest.result);
+              var msg = 'add name in file succeed & file is sending...';
+              debug(msg);
+              if (--numberOfTasks === 0) {
+                transferred();
+              }
+            };
+
+            getRequest.onerror = function() {
+              defaultAdapter.sendFile(targetDevice.address, blob);
+              var msg = 'getFile failed after added name succeed so that ' +
+                        'blob is sending without filename ' +
+                        getRequest.error && getRequest.error.name;
+              debug(msg);
```

Comment:
Is it actually possible to send a blob without a filename? If it is, then why don't we just do that and not bother with the temporary file?  And if it is not possible to do that, then why even call sendFile() if you know it is going to fail?